File: tests/negatives/mcp_client/python/detect-mcp-client.py
# NEGATIVE TEST CASES - Should NOT be detected by the MCP client rule
# These are generic patterns that might be confused with MCP but are not MCP-specific

import asyncio
import json
import os
from typing import Dict, List, Any
from urllib.parse import parse_qs, urlparse

# STANDARD HTTP CLIENT LIBRARIES - should NOT match
import requests
import httpx
import urllib.request
import aiohttp
from requests import Session

# STANDARD ASYNC LIBRARIES - should NOT match
import websockets
import asyncio
import uvloop

# DATABASE CLIENTS - should NOT match
import sqlite3
import psycopg2
from sqlalchemy import create_engine
from databases import Database
import pymongo
from redis import Redis

# GENERIC API CLIENTS - should NOT match
class APIClient:
    def __init__(self, base_url: str):
        self.base_url = base_url
        self.session = requests.Session()

    def list_tools(self):
        """This is NOT MCP - just a generic API method"""
        return self.session.get(f"{self.base_url}/tools")

    def call_tool(self, name: str, args: Dict):
        """This is NOT MCP - just a generic API method"""
        return self.session.post(f"{self.base_url}/tools/{name}", json=args)

    def list_resources(self):
        """This is NOT MCP - just a generic API method"""
        return self.session.get(f"{self.base_url}/resources")

    def read_resource(self, uri: str):
        """This is NOT MCP - just a generic API method"""
        return self.session.get(f"{self.base_url}/resource?uri={uri}")

# GENERIC CLIENT SESSION PATTERNS - should NOT match
class ClientSession:
    """This is NOT MCP ClientSession - just a generic class with same name"""
    def __init__(self, host: str, port: int):
        self.host = host
        self.port = port

    async def connect(self):
        # Generic connection, not MCP
        pass

    async def initialize(self):
        # Generic initialization, not MCP
        pass

    async def list_tools(self):
        # Generic tool listing, not MCP
        return []

# GENERIC ASYNC CONTEXT MANAGERS - should NOT match
async def generic_client_example():
    async with httpx.AsyncClient() as client:
        response = await client.get("https://api.example.com/data")
        return response.json()

async def websocket_client_example():
    async with websockets.connect("ws://localhost:8765") as websocket:
        await websocket.send("hello")
        response = await websocket.recv()

async def database_client_example():
    database = Database("postgresql://user:pass@localhost/db")
    async with database.transaction():
        await database.execute("SELECT * FROM users")

# REGULAR IMPORTS THAT COULD BE CONFUSED - should NOT match
from client import Session  # Not MCP
from api.client import HTTPClient  # Not MCP
import client_lib  # Not MCP
from myapp.clients.stdio import stdio_connection  # Not MCP, similar name

# GENERIC PARAMETER CLASSES - should NOT match
class ServerParameters:
    """Generic server params, not MCP StdioServerParameters"""
    def __init__(self, command: str, args: List[str]):
        self.command = command
        self.args = args

class ConnectionParams:
    """Generic connection params"""
    def __init__(self, host: str, port: int):
        self.host = host
        self.port = port

# GENERIC SERVICE CLIENTS - should NOT match
class ServiceClient:
    def __init__(self, endpoint: str):
        self.endpoint = endpoint

    async def call_service(self, method: str, params: Dict):
        # Generic service call, not MCP
        async with httpx.AsyncClient() as client:
            return await client.post(f"{self.endpoint}/{method}", json=params)

# RPC CLIENTS - should NOT match
import xmlrpc.client
import jsonrpc_requests

def rpc_client_example():
    # XML-RPC client
    proxy = xmlrpc.client.ServerProxy("http://localhost:8000/")
    result = proxy.list_tools()  # NOT MCP

    # JSON-RPC client
    client = jsonrpc_requests.HTTPClient("http://localhost:8001/rpc")
    response = client.call("list_resources")  # NOT MCP

# GRPC CLIENTS - should NOT match
import grpc

async def grpc_client_example():
    async with grpc.aio.insecure_channel("localhost:50051") as channel:
        # Generic gRPC usage, not MCP
        pass

# GENERIC TOOLS AND RESOURCES - should NOT match
def process_tools(tool_list: List[str]):
    """Process a list of tools - NOT MCP related"""
    for tool in tool_list:
        print(f"Processing tool: {tool}")

def read_resource_file(filepath: str):
    """Read a resource file - NOT MCP related"""
    with open(filepath, 'r') as f:
        return f.read()

class ResourceManager:
    """Generic resource manager - NOT MCP"""
    def list_resources(self):
        return os.listdir("./resources")

    def read_resource(self, name: str):
        with open(f"./resources/{name}", 'r') as f:
            return f.read()

# GENERIC ASYNC PATTERNS - should NOT match
async def async_operations():
    # Generic async operations that might look similar
    session = aiohttp.ClientSession()

    # List tools from a generic API
    async with session.get("http://api.example.com/tools") as response:
        tools = await response.json()

    # Call a generic API endpoint
    async with session.post("http://api.example.com/execute", json={"tool": "calculator"}) as response:
        result = await response.json()

    # Read from a generic resource API
    async with session.get("http://api.example.com/resources/readme") as response:
        content = await response.text()

    await session.close()

# GENERIC COMPLETION APIs - should NOT match
from openai import OpenAI
import anthropic

def ai_completion_examples():
    # OpenAI completion - NOT MCP
    openai_client = OpenAI()
    response = openai_client.completions.create(
        model="gpt-3.5-turbo-instruct",
        prompt="Complete this text"
    )

    # Anthropic completion - NOT MCP
    anthropic_client = anthropic.Anthropic()
    message = anthropic_client.messages.create(
        model="claude-3-sonnet-20240229",
        messages=[{"role": "user", "content": "Hello"}]
    )

# GENERIC OAUTH PATTERNS - should NOT match
from requests_oauthlib import OAuth2Session
import oauth2

def oauth_examples():
    # Generic OAuth, not MCP-specific
    oauth = OAuth2Session(client_id="abc123")
    authorization_url, state = oauth.authorization_url("https://provider.com/oauth/authorize")

# CONFIGURATION PATTERNS - should NOT match
config = {
    "server_url": "https://api.example.com",
    "timeout": 30,
    "max_retries": 3,
    "api_key": "secret"
}

# Generic client with similar method names
class GenericClient:
    def __init__(self, config: Dict):
        self.config = config

    def initialize(self):
        # Generic initialization, not MCP
        pass

    def list_tools(self):
        # Generic method, not MCP
        return ["hammer", "screwdriver", "wrench"]

    def call_tool(self, name: str, args: Dict):
        # Generic method, not MCP
        return f"Called {name} with {args}"

# FILE OPERATIONS - should NOT match
def file_operations():
    # Regular file operations that might look like resource operations
    with open("resource.txt", "r") as f:
        content = f.read()

    # List files (not MCP resources)
    import glob
    files = glob.glob("*.txt")

    # Read configuration
    with open("config.json", "r") as f:
        config = json.load(f)

# GENERIC SESSION MANAGEMENT - should NOT match
class Session:
    """Generic session class, not MCP"""
    def __init__(self, connection):
        self.connection = connection

    async def start(self):
        await self.connection.connect()

    async def close(self):
        await self.connection.disconnect()

# TESTING FRAMEWORKS - should NOT match
import pytest
import unittest

class TestClient:
    """Test client - NOT MCP"""
    def list_tools(self):
        return ["test_tool1", "test_tool2"]

# SUBPROCESS USAGE - should NOT match
import subprocess

def run_command():
    # Running subprocesses, not MCP stdio
    result = subprocess.run(["python", "script.py"], capture_output=True, text=True)
    return result.stdout

# REGULAR ASYNC CONTEXT MANAGERS WITH SIMILAR SIGNATURES - should NOT match
async def similar_patterns():
    # Database transaction
    async with Database("sqlite:///db.sqlite").transaction() as transaction:
        await transaction.execute("SELECT 1")

    # HTTP session
    async with httpx.AsyncClient() as (client):
        response = await client.get("http://example.com")

# LOGGING AND MONITORING - should NOT match
import logging

logger = logging.getLogger(__name__)

def log_operations():
    logger.info("Starting operation")
    logger.debug("Debug information")
    logger.warning("Warning message")
    logger.error("Error occurred")

# GENERIC NOTIFICATION SYSTEMS - should NOT match
class NotificationManager:
    async def send_notification(self, message: str):
        # Generic notification, not MCP
        print(f"Notification: {message}")

    async def broadcast_update(self, resource_id: str):
        # Generic broadcast, not MCP
        print(f"Resource {resource_id} updated")

if __name__ == "__main__":
    # Run generic examples that should NOT trigger MCP detection
    client = APIClient("https://api.example.com")
    tools = client.list_tools()

    generic_client = GenericClient(config)
    generic_client.initialize()

    asyncio.run(generic_client_example())
    asyncio.run(async_operations())

File: tests/negatives/mcp_client/typescript/detect-mcp-client.ts
// Negative test cases for TypeScript MCP client detection - these should NOT trigger the rule

// Test case 1: Regular HTTP client imports
import axios from 'axios';
import fetch from 'node-fetch';
import { HttpClient } from '@angular/common/http';
import { Client as GraphQLClient } from 'graphql-request';

// Test case 2: Regular WebSocket clients
import WebSocket from 'ws';
import { io, Socket } from 'socket.io-client';

// Test case 3: Regular database clients
import { MongoClient } from 'mongodb';
import { Client as PgClient } from 'pg';
import { createClient } from 'redis';

// Test case 4: Regular API clients
import { TwitterApi } from 'twitter-api-v2';
import { OpenAI } from 'openai';
import { Anthropic } from '@anthropic-ai/sdk';

// Test case 5: Regular HTTP client class
class HttpApiClient {
  private baseUrl: string;
  private client: any;

  constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
    this.client = axios.create({ baseURL: baseUrl });
  }

  async connect(): Promise<void> {
    // Regular HTTP connection check
    await this.client.get('/health');
  }

  async close(): Promise<void> {
    // No specific close needed for HTTP
  }

  async listItems(): Promise<any[]> {
    const response = await this.client.get('/items');
    return response.data;
  }

  async callEndpoint(endpoint: string, data: any): Promise<any> {
    const response = await this.client.post(`/${endpoint}`, data);
    return response.data;
  }

  async getResource(id: string): Promise<any> {
    const response = await this.client.get(`/resources/${id}`);
    return response.data;
  }
}

// Test case 6: Regular REST API client
class RestApiClient {
  private baseUrl: string;

  constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  async listUsers(): Promise<any[]> {
    const response = await fetch(`${this.baseUrl}/users`);
    return response.json();
  }

  async callApi(endpoint: string, method: string, body?: any): Promise<any> {
    const response = await fetch(`${this.baseUrl}/${endpoint}`, {
      method,
      body: body ? JSON.stringify(body) : undefined,
      headers: { 'Content-Type': 'application/json' }
    });
    return response.json();
  }

  async getDocument(id: string): Promise<any> {
    const response = await fetch(`${this.baseUrl}/documents/${id}`);
    return response.json();
  }
}

// Test case 7: GraphQL client
class GraphQLApiClient {
  private client: GraphQLClient;

  constructor(endpoint: string) {
    this.client = new GraphQLClient(endpoint);
  }

  async listProjects(): Promise<any> {
    const query = `
      query {
        projects {
          id
          name
        }
      }
    `;
    return this.client.request(query);
  }

  async callQuery(query: string, variables?: any): Promise<any> {
    return this.client.request(query, variables);
  }
}

// Test case 8: WebSocket client
class WebSocketClient {
  private ws: WebSocket | null = null;

  async connect(url: string): Promise<void> {
    this.ws = new WebSocket(url);
    return new Promise((resolve) => {
      this.ws!.on('open', resolve);
    });
  }

  async close(): Promise<void> {
    if (this.ws) {
      this.ws.close();
    }
  }

  async listChannels(): Promise<any> {
    return new Promise((resolve) => {
      this.ws!.send(JSON.stringify({ type: 'list_channels' }));
      this.ws!.on('message', (data) => {
        resolve(JSON.parse(data.toString()));
      });
    });
  }

  async callMethod(method: string, params: any): Promise<any> {
    return new Promise((resolve) => {
      this.ws!.send(JSON.stringify({ type: method, params }));
      this.ws!.on('message', (data) => {
        resolve(JSON.parse(data.toString()));
      });
    });
  }
}

// Test case 9: Socket.IO client
class SocketIOClient {
  private socket: Socket;

  constructor(url: string) {
    this.socket = io(url);
  }

  async connect(): Promise<void> {
    return new Promise((resolve) => {
      this.socket.on('connect', resolve);
    });
  }

  async close(): Promise<void> {
    this.socket.disconnect();
  }

  async listRooms(): Promise<any> {
    return new Promise((resolve) => {
      this.socket.emit('list_rooms', (response: any) => {
        resolve(response);
      });
    });
  }

  async callEvent(event: string, data: any): Promise<any> {
    return new Promise((resolve) => {
      this.socket.emit(event, data, (response: any) => {
        resolve(response);
      });
    });
  }
}

// Test case 10: Database client
class DatabaseClient {
  private client: MongoClient;

  constructor(connectionString: string) {
    this.client = new MongoClient(connectionString);
  }

  async connect(): Promise<void> {
    await this.client.connect();
  }

  async close(): Promise<void> {
    await this.client.close();
  }

  async listCollections(): Promise<any> {
    const db = this.client.db();
    return db.listCollections().toArray();
  }

  async callProcedure(name: string, params: any): Promise<any> {
    const db = this.client.db();
    return db.collection('procedures').findOne({ name, params });
  }

  async getDocument(collection: string, id: string): Promise<any> {
    const db = this.client.db();
    return db.collection(collection).findOne({ _id: id });
  }
}

// Test case 11: Redis client
class CacheClient {
  private client: any;

  constructor() {
    this.client = createClient();
  }

  async connect(): Promise<void> {
    await this.client.connect();
  }

  async close(): Promise<void> {
    await this.client.quit();
  }

  async listKeys(): Promise<string[]> {
    return this.client.keys('*');
  }

  async callCommand(command: string, ...args: any[]): Promise<any> {
    return this.client.sendCommand([command, ...args]);
  }

  async getValue(key: string): Promise<any> {
    return this.client.get(key);
  }
}

// Test case 12: OpenAI client (similar AI client but not MCP)
class OpenAIClient {
  private client: OpenAI;

  constructor(apiKey: string) {
    this.client = new OpenAI({ apiKey });
  }

  async listModels(): Promise<any> {
    return this.client.models.list();
  }

  async callCompletion(prompt: string): Promise<any> {
    return this.client.completions.create({
      model: 'text-davinci-003',
      prompt,
      max_tokens: 100
    });
  }

  async getEmbedding(text: string): Promise<any> {
    return this.client.embeddings.create({
      model: 'text-embedding-ada-002',
      input: text
    });
  }
}

// Test case 13: Anthropic client (similar AI client but not MCP)
class AnthropicClient {
  private client: Anthropic;

  constructor(apiKey: string) {
    this.client = new Anthropic({ apiKey });
  }

  async listModels(): Promise<any> {
    // Anthropic doesn't have a public models endpoint
    return ['claude-3-opus-20240229', 'claude-3-sonnet-20240229'];
  }

  async callMessage(content: string): Promise<any> {
    return this.client.messages.create({
      model: 'claude-3-opus-20240229',
      max_tokens: 1000,
      messages: [{ role: 'user', content }]
    });
  }

  async getResponse(prompt: string): Promise<any> {
    return this.client.messages.create({
      model: 'claude-3-sonnet-20240229',
      max_tokens: 500,
      messages: [{ role: 'user', content: prompt }]
    });
  }
}

// Test case 14: Generic API client with similar method names
class GenericApiClient {
  private baseUrl: string;
  private httpClient: any;

  constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
    this.httpClient = axios.create({ baseURL: baseUrl });
  }

  async connect(): Promise<void> {
    // Generic connection test
    await this.httpClient.get('/ping');
  }

  async close(): Promise<void> {
    // Generic cleanup
  }

  // These method names are similar to MCP but in different context
  async listTools(): Promise<any> {
    // Regular API endpoint for development tools
    const response = await this.httpClient.get('/dev/tools');
    return response.data;
  }

  async callTool(toolName: string, params: any): Promise<any> {
    // Regular API call to a development tool endpoint
    const response = await this.httpClient.post(`/dev/tools/${toolName}`, params);
    return response.data;
  }

  async listResources(): Promise<any> {
    // Regular API endpoint for cloud resources
    const response = await this.httpClient.get('/cloud/resources');
    return response.data;
  }

  async readResource(resourceId: string): Promise<any> {
    // Regular API call to read cloud resource
    const response = await this.httpClient.get(`/cloud/resources/${resourceId}`);
    return response.data;
  }

  async listPrompts(): Promise<any> {
    // Regular API endpoint for UI prompts/templates
    const response = await this.httpClient.get('/ui/prompts');
    return response.data;
  }

  async getPrompt(promptId: string): Promise<any> {
    // Regular API call to get UI prompt
    const response = await this.httpClient.get(`/ui/prompts/${promptId}`);
    return response.data;
  }
}

// Test case 15: Message queue client
class MessageQueueClient {
  private connection: any;

  async connect(url: string): Promise<void> {
    // Generic message queue connection
  }

  async close(): Promise<void> {
    // Generic cleanup
  }

  async request(message: any, schema?: any): Promise<any> {
    // Generic message queue request
    return { status: 'processed', data: message };
  }
}

// Test case 16: Configuration objects that might look similar
const apiConfig = {
  capabilities: {
    caching: true,
    compression: true,
    retries: 3
  }
};

const serviceConfig = {
  capabilities: {
    monitoring: {},
    logging: {},
    metrics: {}
  }
};

// Test case 17: Class with private client property (generic)
class ServiceManager {
  private client: any;
  private httpClient: any;

  constructor() {
    this.client = axios.create();
    this.httpClient = fetch;
  }
}

// Test case 18: Transport-like classes (but not MCP)
class HttpTransport {
  constructor(options: { url: string }) {
    // Generic HTTP transport
  }
}

class WebSocketTransport {
  constructor(url: string) {
    // Generic WebSocket transport
  }
}

// Test case 19: Child process management (not MCP stdio)
import { spawn } from 'child_process';

class ProcessManager {
  async startProcess(command: string, args: string[]): Promise<void> {
    const child = spawn(command, args);
    // This is generic process management, not MCP
  }
}

// Test case 20: Regular schema definitions
const ApiSchema = {
  type: 'object',
  properties: {
    result: { type: 'string' }
  }
};

const ResponseSchema = {
  type: 'object',
  properties: {
    data: { type: 'array' },
    status: { type: 'string' }
  }
};

// Export for potential use
export {
  HttpApiClient,
  RestApiClient,
  GraphQLApiClient,
  WebSocketClient,
  SocketIOClient,
  DatabaseClient,
  CacheClient,
  OpenAIClient,
  AnthropicClient,
  GenericApiClient,
  MessageQueueClient,
  ServiceManager,
  ProcessManager
};